### DISCUSSION TITLE
2023-12-22 - Co-creation patterns - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by bradystroud** [SSWConsulting/SSW.Rules.Content#7632] [Chewing the Fat]
Hey SSWers,

Lets talk about co-creating!

https://www.ssw.com.au/rules/co-creation-patterns/

**Comment by jakebayliss** [comment]
Definitely think Pair-Programming has its benefits but shouldn't be the default. My view is it is very dependent on the PBI and should be used for the more complex problems/solutions

**Comment by MarcusIrmscherSSW** [comment]
It seemy like a good oportunity to pick up different ways of doing the same task. Often people will have different solutions for the same problem👍🃏🃏

**Comment by 2fernandez** [comment]
Co-creation works, as long as you understand people's strengths, weaknesses and experiences.
Leveraging people's strengths in the team will be far more effective than trying to conform or fight against them.

**Reply by JudyMao99** [reply]
It is a good chance to learn more coding skills!

**Comment by bettybondoc** [comment]
Co-creation has a lot benefits and would work well even for non-devs. I've had other designers and devs sit with me to do a live edit of a design or a website mockup in the past. I find this approach especially beneficial when dealing with ambiguous feedback or when we need to choose from several potential layouts. The real-time collaboration and exchange of ideas streamline the decision-making process and enhance the final outcome.

**Comment by apforsyth** [comment]
In video production we use "edit reviews" in a similar way, but also work solo for significant periods

**Comment by BenDunkerleySSW** [comment]
On the non-coding side of this, the hardest part is often coming to a middle ground on how a task should be approached when you have someone working on it with you. If you are in agreement, its smooth sailing and the pair working helps with efficiency and reducing mistakes. However, if there is a disagreement, you can often lose time trying to discuss the correct solution. So I think it should be reserved for tasks that really need pair working, not a mandatory path for every task.

**Comment by Calinator444** [comment]
Co creation sounds like a good means of ensuring quality. Later in the day when developers become more cognitively overloaded a lot of mistakes might slip through the cracks. For instance variables might be given poor names or bugs might be written into the code.

**Comment by chrisschultzssw** [comment]
The small amount of co-creation I have done has gone well - it will be interesting to think of ways to use it for more SysAdmin tasks.

**Comment by cvionnet** [comment]
Pair Programming is a great option, especially when joining a new project or team. It can help to learn about the project, technology but also people.
The video and rule is doing a great job at giving a clearer definition of the roles of "driver-navigator".

**Comment by TobyChurches** [comment]
Co-Creation and Pair programming is great, but I don't believe that it is required for all tasks. Small or simple tasks are better developed solo for better usage of dev resources.
Saying that, for larger PBI's or when I was first starting out, the TimePro Team has always been a pleasure to work with 😁

**Reply by SamBWagner** [reply]
#ALLTHELOVEFORMYMANTOBY ❤️

Pairing up was always fun with you my g

**Comment by isaaclombardssw** [comment]
It's a good strategy, when I've done projects alongside those more familiar with a language/tool it has helped me get an idea of the conventions and thought process of it. Also practical as means of QA.

**Comment by leahy268** [comment]
I have done a small amount of pair programming which has been great to teach me new techniques.  It really couldn't always be used for what I am mostly doing (GitHub actions and bicep) as most of the time the other person would be watching while I make sure all the required parameters are there and then deploying/waiting and hoping that I got everything correct.  

However it has been super useful for making large changes and verifying everything was correct to the best of our knowledge before deployment.

Also with SysAdmin work, having someone look over your shoulder to either understand a problem or to watch you make a major change is great.

**Comment by JackDevAU** [comment]
Not useful for everything but for tough problems 2 heads are better than 1. Good video and 2 rules - will look into the VS Code mob extension 👀 very cool

**Comment by jeoffreyfischer** [comment]
The driver/navigator is a great concept! It benefits both developers in their learning

**Comment by YalalCakirSSW** [comment]
In theory it seems like a perfect idea. But I'm not sure how sustainable it will be in the long run because I've never tried pair-programming.

For Junior Devs, it's a great way to learn and get feedback instantly. However, in terms of Expert Devs, they are spending way more time on something they could do half of the time. Pros and cons. I am really curious about pair-programming. I hope I can get a good chance to experience it.

**Comment by synnoveisabel** [comment]
I really like pair programming - it's a cool way to learn stuff and get feedback on the fly

**Comment by Ben0189** [comment]
Pair programming is an effective method for tackling complex problems. To enhance the process and save time, preparation is key. This includes being clear on the PBI to work on, crafting a draft solution or come up with a simple solution idea before pair programming, and reviewing relevant documentation.

**Comment by amankumarrr** [comment]
The concept of having a Driver and Navigator is beneficial; it significantly enhanced the pair programming experience.

**Comment by AttackOnMorty** [comment]
Pair Programming with TDD.

<img width="500" alt="image" src="https://github.com/SSWConsulting/SSW.Rules.Content/assets/37203901/79a8d59c-17b5-4940-9449-9d826f3e3327">

**Comment by ncn-ssw** [comment]
If I was forced to program in the manner described in the 1st 5 minutes of the video, I would resign. The absolute last thing I want is to be interrupted as I'm resolving a non-trivial issue, as writing the code is itself part of my thinking process. Pair reviewing, by contrast, can be invaluable for non-trivial PRs.

**Reply by SamBWagner** [reply]
I think that's totally valid. I do think there's something to be said that there's a paired contract between the two devs to work in the way that makes sense. I.e. if you don't like being interrupted when writing a variable name or setting up a class but would rather finish an idea then just ask "what do you think?" to your pair then that's totally the way to go.

People over Process! :)

Defs not a hard and fast rule though

**Comment by AntPolkanov** [comment]
Pair programming is great. 
Here are some thoughts from my experience:

-  💻It's better if the driver uses their own laptop, because every developer tweaks their IDEs for themselves. Even such a small thing like different keyboard and mouse can cause a lot of discomfort. 
-  🤓At least one of the developers should know the technology and be comfortable with the code base. Otherwise, pair programming comes down to just 2 people reading docs and poking around.
- 🥱It's emotionally draining - after a few hours you feel exhausted (that's why the author of the video mentioned it several times). Better to start small.

**Comment by tomek-i** [comment]
While it sounds great, everything comes at a cost. It certainly has a lot of advantages. Most of the times we are restricted by time and budget of the client and need to skip on all the good things that would generate good and robust code. 

For internal work we should definitely try them out more often and get more efficient with it.

**Comment by ricksu978** [comment]
Pair programming is valuable for complex tasks and learning, but it's important to balance it with individual work based on task simplicity and developer expertise.

**Comment by tkapa** [comment]
I think that Pair Programming almost always results in a positive outcome. 

I pair programmed the Co Creation rule with @SamBWagner to correct an issue with the Driver/Navigator roles.
It worked well!

**Comment by CoffeeBugHub** [comment]
Great rules, if PO encourage it and we should try more.

**Comment by christoment** [comment]
Sounds good on paper, but I think this should only be doing sparingly on high risk PBI.

My opinion:

1. Pair programming is **_very_** time consuming - IDE problems (intellisense, refactor capability, etc), difficult access to debug codes, code can only run in host, and overall discomfort in coding in this environment.
2. Overall lower code throughput with the trade-off of having on-the-fly code review.
3. It can be a tool for mentoring, but come with a very high cost of doubling the developer count - really hard to sell this to a client
4. Cognitive overload - pair programming is mentally exhausting since you have to maintain constant focus with less autonomy

**Reply by SamBWagner** [reply]
I've personally not found the pain in time consumption to be that bad. I worked for 5+ months with another SSW dev at CER and we exclusively pair programmed. It was a little hard at first but we found a groove and really started cooking! 

The tooling for it is definitely lacking and it's a pretty poor showing from most IDE creators/maintainers that they're not better.

It's not for everyone, not every problem that's for sure.

**Comment by calumjs** [comment]
The video discussed the benefits of pairing up beginners with experts for mentoring. But there's another big plus that wasn't mentioned: bringing someone new to the field into the mix. Fresh eyes can spot problems that veterans might ignore, and can potentially come up with new ways to tackle old challenges. It's kind of like the idea of 'Shoshin' or a 'beginner's mind' in Zen Buddhism, which is about being open and eager without being bogged down by past ideas. This approach doesn't just help the newbies learn; it can also breathe new life into how we solve problems.

**Comment by KostaMadorsky** [comment]
Pair programming is great when design for a PBI is not clear. A good way to mentor or get mentored. 
Saves some time on reviewing and reworking later on.

Still would not be my default way of doing things.

**Comment by SamBWagner** [comment]
I worked EXCLUSIVELY in a pair programming setup for ~5 months and found it to be great! I think there's something to be said for being able to work individually but being unable to work collaboratively is a big red-flag in my book.

I think that a lot of the time, people try it and then give up because it's hard initially. It's a skill to be developed though and you need to invest the time in it. First impressions are pretty well always going to be wrong in this case.

It takes time and practice.

**Comment by Harry-Ross** [comment]
Super interesting concept of having a driver and an instructor for doing pair programming - I've found it to be beneficial in the rare times I have done it. However, I've often found it does slow down the process compared to focusing in on a certain problem on your own.

**Comment by ozair-ashfaque** [comment]
I believe that pair programming is effective for tackling complex problems. But it is time-consuming and costly, the decision to use pair programming should be based on the PBI.

**Comment by MattParkerDev** [comment]
Pair programming can be effective when used correctly. I think its a fine line between synergy and codependence, and my hot take is that someone else should still review your PR after you finish.

![NvoA](https://github.com/SSWConsulting/SSW.Rules.Content/assets/61717342/a81d921b-21dd-48ca-926d-58deccf6005e)

**Comment by PennyWalker** [comment]
Pair programming looks like it would be the perfect way to solve a complex problem - you can quickly work through it with someone and you don't get blocked on a PR. I'm not sure if it would be very efficient for every project or scenario, but I can see the benefits for people who are less experienced and want to learn quickly while coding without any risk of making terrible errors.

**Comment by bahjatalaadelssw** [comment]
I always think 2 minds are better than 1 😊

**Comment by jaydenalchin** [comment]
As a designer / frontend-only CSS wizard, pair programming is an awesome way for me to leverage the deeper skillset of a dev (and appreciate the sheer amount of keyboard shortcuts they have committed to memory) when I might otherwise spend a lot more time figuring things out myself.

**Comment by AlexRothwell** [comment]
I haven't done much pair programming but am often happy to take the role of a navigator. Personally, I struggle to listen and type at the same time so when acting as the driver I would definitely need to have an agreement in place like others have mentioned. I think having discussions only occur between coding would work well

**Comment by drwharris** [comment]
Pair coding, designing, etc 100% but not sure I agree that co creators should be approving their PR. One of the points of PR review is to get independent input into what you are proposing.

**Comment by jerryluossw** [comment]
I think not all PBIs are suitable for pair programming, in some cases, the pair programming experience is great.

**Comment by jimmidier** [comment]
Great rule. The main purpose of co working is to find blind spots of the other

**Comment by Levijj22** [comment]
It can be difficult to sell if the PO only wants 1 dev, especially since twice the dev doesn't necessarily mean twice the speed.
The key is to sell, not twice the speed, but double the quality.

**Reply by MattParkerDev** [reply]
Twice the pride, double the fall

**Comment by danielmackay** [comment]
There are some great takeaways from this rule, particular about how to convince the client of the value of co-creation.

The tips on HOW to pair-program are great.  Especially, about the navigator concentrating on the big picture and the driver concentrating on the implementation details.  

⚠️ It's important to keep in mind that micromanaging every keystroke is no fun for anyone.

**Comment by babakamyljanovssw** [comment]
I didn't have much experience on pair-programming , but I think it depends on the PBI and not every task can be suitable for it. I think for sure it has benefits, junior devs can get a lot from this experience and improve their skills.

**Comment by KristenHu** [comment]
Two heads are better than one, but I don't think that all tasks and PBIs are suitable for pair-programming, it's not efficient sometimes.

**Comment by PothieuG** [comment]
Very exhaustive rule !
Of course pair-programming is the must when you are working, but it's hard to convince the product owner about it...

**Comment by camillars** [comment]
Co-creation is awesome as you get the chance to figure out the problem and fix it faster, but also learn along the way. However, I agree with most that it is required for all tasks/projects. Working in pairs can be great, but also more complicated or inefficient depending on the task.

**Comment by sylhuang** [comment]
I like the concept of pair programming, but I just can't concentrate while someone is looking at my screen... 😅

**Comment by matt-goldman** [comment]
Pair programming is great. It's a tool and just like any other you have to know when and how to use it. Defaulting to pair programming for all our work all the time is low value. It costs double to get any work done.

**Comment by AnkyChe** [comment]
I love the concept of pair programming. For some complex financial work, it's also good to work in pairs. However, I think pair programming is not suitable for all tasks.

**Comment by SofieHong** [comment]
Great rule 🌟 I agree that 2 minds are better than 1 (Like @bahjatalaadelssw said)
During video production, "Video Review" is like doing pair programming.

**Comment by Aibono1225** [comment]
I can see the value of pair programming, but personally I still prefer to work alone to better concentrate on my task

**Comment by BrookJeynes** [comment]
Pair programming is a useful tool. It shouldn't always be reached for first, but it's a nice to have in the toolbox.

**Comment by vladislav-kir** [comment]
Pair programming is useful, especially for the harder PBIs. However, for most tasks I prefer to concentrate on them alone.

The rules were great - informative but easy to follow :)

**Comment by steven0x51** [comment]
Pair programming is useful in some case, but not for every coding task.

**Comment by tino-liu** [comment]
I love the outcome pair programming yields. But it is very draining during the process.

**Comment by sethdaily** [comment]
(transplanting this idea from coding to writing/designing)
It's a useful technique when used wisely.

**Comment by jackreimers** [comment]
I think pair programming can be very useful but there is also value in developing your own investigation and problem solving skills where time permits.

**Comment by Marxoz** [comment]
I love pair programming. It is not only good for building quality features, but also good for building relationships with your team members. Always try and make it fun :)

**Comment by stephanfakossw** [comment]
I had very good experience with co-creation. Live feedback and improvement, helps everyone in team feel motivated which helps the project move along.
